File: data/recomposed_projects/gpt-4o-2024-11-20/body/0.0/commons-fileupload/src/main/org/apache/commons/fileupload/util/mime/MimeUtility.py
from __future__ import annotations
import re
import os
from io import BytesIO
import io
import typing
from typing import *
from src.main.org.apache.commons.fileupload.util.mime.Base64Decoder import *
from src.main.org.apache.commons.fileupload.util.mime.ParseException import *
from src.main.org.apache.commons.fileupload.util.mime.QuotedPrintableDecoder import *


class MimeUtility:

    __MIME2JAVA: typing.Dict[str, str] = {}

    __LINEAR_WHITESPACE: str = " \t\r\n"
    __ENCODED_TOKEN_FINISHER: str = "?="
    __ENCODED_TOKEN_MARKER: str = "=?"
    __QUOTEDPRINTABLE_ENCODING_MARKER: str = "Q"
    __BASE64_ENCODING_MARKER: str = "B"
    __US_ASCII_CHARSET: str = "US-ASCII"

    @staticmethod
    def run_static_init():
        MimeUtility.__MIME2JAVA["iso-2022-cn"] = "ISO2022CN"
        MimeUtility.__MIME2JAVA["iso-2022-kr"] = "ISO2022KR"
        MimeUtility.__MIME2JAVA["utf-8"] = "UTF8"
        MimeUtility.__MIME2JAVA["utf8"] = "UTF8"
        MimeUtility.__MIME2JAVA["ja_jp.iso2022-7"] = "ISO2022JP"
        MimeUtility.__MIME2JAVA["ja_jp.eucjp"] = "EUCJIS"
        MimeUtility.__MIME2JAVA["euc-kr"] = "KSC5601"
        MimeUtility.__MIME2JAVA["euckr"] = "KSC5601"
        MimeUtility.__MIME2JAVA["us-ascii"] = "ISO-8859-1"
        MimeUtility.__MIME2JAVA["x-us-ascii"] = "ISO-8859-1"

    @staticmethod
    def decodeText(text: str) -> str:
        if MimeUtility.__ENCODED_TOKEN_MARKER not in text:
            return text

        offset = 0
        end_offset = len(text)

        start_white_space = -1
        end_white_space = -1

        decoded_text = []

        previous_token_encoded = False

        while offset < end_offset:
            ch = text[offset]

            if ch in MimeUtility.__LINEAR_WHITESPACE:  # whitespace found
                start_white_space = offset
                while offset < end_offset:
                    ch = text[offset]
                    if ch in MimeUtility.__LINEAR_WHITESPACE:  # whitespace found
                        offset += 1
                    else:
                        end_white_space = offset
                        break
            else:
                word_start = offset

                while offset < end_offset:
                    ch = text[offset]
                    if ch not in MimeUtility.__LINEAR_WHITESPACE:  # not white space
                        offset += 1
                    else:
                        break

                word = text[word_start:offset]
                if word.startswith(MimeUtility.__ENCODED_TOKEN_MARKER):
                    try:
                        decoded_word = MimeUtility.__decodeWord(word)

                        if not previous_token_encoded and start_white_space != -1:
                            decoded_text.append(text[start_white_space:end_white_space])
                            start_white_space = -1

                        previous_token_encoded = True
                        decoded_text.append(decoded_word)
                        continue

                    except ParseException:
                        pass

                if start_white_space != -1:
                    decoded_text.append(text[start_white_space:end_white_space])
                    start_white_space = -1

                previous_token_encoded = False
                decoded_text.append(word)

        return "".join(decoded_text)

    @staticmethod
    def __javaCharset(charset: str) -> str:
        if charset is None:
            return None

        mapped_charset = MimeUtility.__MIME2JAVA.get(charset.lower())
        if mapped_charset is None:
            return charset
        return mapped_charset

    @staticmethod
    def __decodeWord(word: str) -> str:
        if not word.startswith(MimeUtility.__ENCODED_TOKEN_MARKER):
            raise ParseException(f"Invalid RFC 2047 encoded-word: {word}")

        charset_pos = word.find("?", 2)
        if charset_pos == -1:
            raise ParseException(f"Missing charset in RFC 2047 encoded-word: {word}")

        charset = word[2:charset_pos].lower()

        encoding_pos = word.find("?", charset_pos + 1)
        if encoding_pos == -1:
            raise ParseException(f"Missing encoding in RFC 2047 encoded-word: {word}")

        encoding = word[charset_pos + 1 : encoding_pos]

        encoded_text_pos = word.find(
            MimeUtility.__ENCODED_TOKEN_FINISHER, encoding_pos + 1
        )
        if encoded_text_pos == -1:
            raise ParseException(
                f"Missing encoded text in RFC 2047 encoded-word: {word}"
            )

        encoded_text = word[encoding_pos + 1 : encoded_text_pos]

        if len(encoded_text) == 0:
            return ""

        try:
            out = io.BytesIO()

            encoded_data = encoded_text.encode(MimeUtility.__US_ASCII_CHARSET)

            if encoding == MimeUtility.__BASE64_ENCODING_MARKER:
                Base64Decoder.decode(encoded_data, out)
            elif encoding == MimeUtility.__QUOTEDPRINTABLE_ENCODING_MARKER:
                QuotedPrintableDecoder.decode(encoded_data, out)
            else:
                raise ValueError(f"Unknown RFC 2047 encoding: {encoding}")

            decoded_data = out.getvalue()
            return decoded_data.decode(MimeUtility.__javaCharset(charset))
        except Exception as e:
            raise ValueError("Invalid RFC 2047 encoding") from e

    def __init__(self) -> None:
        raise NotImplementedError("This class cannot be instantiated.")


MimeUtility.run_static_init()

File: logs/glue/commons-fileupload/src/main/java/org/apache/commons/fileupload/util/mime/MimeUtility.java
package org.apache.commons.fileupload.util.mime;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import org.apache.commons.fileupload.ContextInitializer;
import org.apache.commons.fileupload.ExceptionHandler;
import org.apache.commons.fileupload.IntegrationUtils;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

public final class MimeUtility {
  private transient Value obj =
      IntegrationUtils.createDefaultPythonObject(
          ContextInitializer.getPythonClass(
              "main/org/apache/commons/fileupload/util/mime/MimeUtility.py", "MimeUtility"));
  public static final String US_ASCII_CHARSET = "US-ASCII";
  public static final String BASE64_ENCODING_MARKER = "B";
  public static final String QUOTEDPRINTABLE_ENCODING_MARKER = "Q";
  public static final String ENCODED_TOKEN_MARKER = "=?";
  public static final String ENCODED_TOKEN_FINISHER = "?=";
  public static final String LINEAR_WHITESPACE = " \t\r\n";
  public static final Map<String, String> MIME2JAVA = new HashMap<String, String>();

  static {
    MIME2JAVA.put("iso-2022-cn", "ISO2022CN");
    MIME2JAVA.put("iso-2022-kr", "ISO2022KR");
    MIME2JAVA.put("utf-8", "UTF8");
    MIME2JAVA.put("utf8", "UTF8");
    MIME2JAVA.put("ja_jp.iso2022-7", "ISO2022JP");
    MIME2JAVA.put("ja_jp.eucjp", "EUCJIS");
    MIME2JAVA.put("euc-kr", "KSC5601");
    MIME2JAVA.put("euckr", "KSC5601");
    MIME2JAVA.put("us-ascii", "ISO-8859-1");
    MIME2JAVA.put("x-us-ascii", "ISO-8859-1");
  }

  public Value getPythonObject() {
    return obj;
  }

  public void setPythonObject(Value obj) {
    this.obj = obj;
  }

  public static String decodeText(String text) throws UnsupportedEncodingException {

    try {

      if (text.indexOf(ENCODED_TOKEN_MARKER) < 0) {
        return text;
      }

      int offset = 0;
      int endOffset = text.length();

      int startWhiteSpace = -1;
      int endWhiteSpace = -1;

      StringBuilder decodedText = new StringBuilder(text.length());

      boolean previousTokenEncoded = false;

      while (offset < endOffset) {
        char ch = text.charAt(offset);

        if (LINEAR_WHITESPACE.indexOf(ch) != -1) { // whitespace found
          startWhiteSpace = offset;
          while (offset < endOffset) {
            ch = text.charAt(offset);
            if (LINEAR_WHITESPACE.indexOf(ch) != -1) { // whitespace found
              offset++;
            } else {
              endWhiteSpace = offset;
              break;
            }
          }
        } else {
          int wordStart = offset;

          while (offset < endOffset) {
            ch = text.charAt(offset);
            if (LINEAR_WHITESPACE.indexOf(ch) == -1) { // not white space
              offset++;
            } else {
              break;
            }
          }
          String word = text.substring(wordStart, offset);
          if (word.startsWith(ENCODED_TOKEN_MARKER)) {
            try {
              String decodedWord = decodeWord(word);

              if (!previousTokenEncoded && startWhiteSpace != -1) {
                decodedText.append(text.substring(startWhiteSpace, endWhiteSpace));
                startWhiteSpace = -1;
              }
              previousTokenEncoded = true;
              decodedText.append(decodedWord);
              continue;

            } catch (ParseException e) {
            }
          }
          if (startWhiteSpace != -1) {
            decodedText.append(text.substring(startWhiteSpace, endWhiteSpace));
            startWhiteSpace = -1;
          }
          previousTokenEncoded = false;
          decodedText.append(word);
        }
      }

      return decodedText.toString();

    } catch (Throwable ExceptionObjectForCaching) {
      ExceptionHandler.ERR = ExceptionObjectForCaching;
      throw ExceptionObjectForCaching;
    }
  }

  public static String javaCharset(String charset) {

    if (charset == null) {
      return null;
    }

    String mappedCharset = MIME2JAVA.get(charset.toLowerCase(Locale.ENGLISH));
    if (mappedCharset == null) {
      return charset;
    }
    return mappedCharset;
  }

  public static String decodeWord(String word) throws ParseException, UnsupportedEncodingException {

    try {

      if (!word.startsWith(ENCODED_TOKEN_MARKER)) {
        throw new ParseException("Invalid RFC 2047 encoded-word: " + word);
      }

      int charsetPos = word.indexOf('?', 2);
      if (charsetPos == -1) {
        throw new ParseException("Missing charset in RFC 2047 encoded-word: " + word);
      }

      String charset = word.substring(2, charsetPos).toLowerCase(Locale.ENGLISH);

      int encodingPos = word.indexOf('?', charsetPos + 1);
      if (encodingPos == -1) {
        throw new ParseException("Missing encoding in RFC 2047 encoded-word: " + word);
      }

      String encoding = word.substring(charsetPos + 1, encodingPos);

      int encodedTextPos = word.indexOf(ENCODED_TOKEN_FINISHER, encodingPos + 1);
      if (encodedTextPos == -1) {
        throw new ParseException("Missing encoded text in RFC 2047 encoded-word: " + word);
      }

      String encodedText = word.substring(encodingPos + 1, encodedTextPos);

      if (encodedText.length() == 0) {
        return "";
      }

      try {
        ByteArrayOutputStream out = new ByteArrayOutputStream(encodedText.length());

        byte[] encodedData = encodedText.getBytes(US_ASCII_CHARSET);

        if (encoding.equals(BASE64_ENCODING_MARKER)) {
          Base64Decoder.decode(encodedData, out);
        } else if (encoding.equals(QUOTEDPRINTABLE_ENCODING_MARKER)) { // maybe quoted printable.
          QuotedPrintableDecoder.decode(encodedData, out);
        } else {
          throw new UnsupportedEncodingException("Unknown RFC 2047 encoding: " + encoding);
        }
        byte[] decodedData = out.toByteArray();
        return new String(decodedData, javaCharset(charset));
      } catch (IOException e) {
        throw new UnsupportedEncodingException("Invalid RFC 2047 encoding");
      }

    } catch (Throwable ExceptionObjectForCaching) {
      ExceptionHandler.ERR = ExceptionObjectForCaching;
      throw ExceptionObjectForCaching;
    }
  }

  public MimeUtility() {}

  public java.util.Map pyToJ() {
    java.util.Map idMap = new java.util.HashMap();
    return pyToJ(idMap);
  }

  public java.util.Map pyToJ(java.util.Map idMap) {

    return idMap;
  }

  public Value jToPy() {
    Value idMap = IntegrationUtils.mapToPython(new java.util.HashMap());
    return jToPy(idMap);
  }

  public Value jToPy(Value idMap) {

    this.obj.invokeMember("__setattr__", "javaObj", this);

    return idMap;
  }
}
